File: mobile/lib/features/colorimetry/engine/correction.dart
/// Colour correction from observed reference patches to true printed colour.
///
/// Mirrors `backend/app/services/colorimetry/correction.py`.
///
/// The fit is performed in **linear** RGB, because that is the space in which a
/// camera's response to light is approximately affine. Fitting in gamma-encoded
/// sRGB would fold the transfer function into the fit and give a correction only
/// valid at the exposure it was fitted at.
///
/// Two models are tried and the better *cross-validated* one wins. Leave-one-out
/// matters: the 3×3 model has enough freedom to fit 12 patches well while
/// generalising worse to the strip colour, and an in-sample residual would hide
/// exactly that.
library;

import 'dart:math' as math;

import 'color_math.dart';

class ColorCorrection {
  ColorCorrection({
    required this.model,
    required this.matrix,
    required this.offset,
    required this.residualsDe,
    required this.cvResidualsDe,
    required this.nPatches,
  });

  /// 'per_channel' or 'affine33'.
  final String model;

  /// Row-major 3x3.
  final List<List<double>> matrix;
  final List<double> offset;

  final List<double> residualsDe;
  final List<double> cvResidualsDe;
  final int nPatches;

  double get meanResidualDe =>
      residualsDe.reduce((double a, double b) => a + b) / residualsDe.length;

  double get meanCvResidualDe =>
      cvResidualsDe.reduce((double a, double b) => a + b) / cvResidualsDe.length;

  double get worstCvResidualDe => cvResidualsDe.reduce(math.max);

  List<double> applyLinear(List<double> linearRgb) {
    return <double>[
      for (int r = 0; r < 3; r++)
        matrix[r][0] * linearRgb[0] +
            matrix[r][1] * linearRgb[1] +
            matrix[r][2] * linearRgb[2] +
            offset[r],
    ];
  }

  /// Correct an 8-bit sRGB colour and return it as CIE L*a*b*.
  Lab applyToLab(List<double> rgbU8) {
    final List<double> linear = <double>[
      srgbToLinear(rgbU8[0] / 255.0),
      srgbToLinear(rgbU8[1] / 255.0),
      srgbToLinear(rgbU8[2] / 255.0),
    ];
    final List<double> corrected = applyLinear(linear)
        .map((double v) => v.clamp(0.0, 1.0).toDouble())
        .toList(growable: false);
    return _labOfLinear(corrected);
  }

  Map<String, dynamic> toMap() => <String, dynamic>{
        'model': model,
        'n_patches': nPatches,
        'mean_residual_de': meanResidualDe,
        'mean_cv_residual_de': meanCvResidualDe,
        'worst_cv_residual_de': worstCvResidualDe,
      };
}

Lab _labOfLinear(List<double> linearRgb) {
  final double r = linearToSrgb(linearRgb[0]).clamp(0.0, 1.0).toDouble();
  final double g = linearToSrgb(linearRgb[1]).clamp(0.0, 1.0).toDouble();
  final double b = linearToSrgb(linearRgb[2]).clamp(0.0, 1.0).toDouble();
  return xyzToLab(srgbToXyz(r, g, b));
}

/// Least-squares gain and offset per channel, in linear RGB.
({List<List<double>> matrix, List<double> offset}) _fitPerChannel(
  List<List<double>> observed,
  List<List<double>> expected,
) {
  final List<List<double>> matrix = <List<double>>[
    <double>[1, 0, 0],
    <double>[0, 1, 0],
    <double>[0, 0, 1],
  ];
  final List<double> offset = <double>[0, 0, 0];

  for (int c = 0; c < 3; c++) {
    final int n = observed.length;
    double sumX = 0, sumY = 0, sumXX = 0, sumXY = 0;
    for (int i = 0; i < n; i++) {
      final double x = observed[i][c];
      final double y = expected[i][c];
      sumX += x;
      sumY += y;
      sumXX += x * x;
      sumXY += x * y;
    }
    final double denominator = n * sumXX - sumX * sumX;
    if (denominator.abs() < 1e-12) {
      matrix[c][c] = 1.0;
      offset[c] = 0.0;
      continue;
    }
    final double gain = (n * sumXY - sumX * sumY) / denominator;
    matrix[c][c] = gain;
    offset[c] = (sumY - gain * sumX) / n;
  }
  return (matrix: matrix, offset: offset);
}

/// Least-squares 3x3 matrix plus offset, solved by normal equations.
({List<List<double>> matrix, List<double> offset})? _fitAffine33(
  List<List<double>> observed,
  List<List<double>> expected,
) {
  const int k = 4; // r, g, b, 1
  final List<List<double>> ata =
      List<List<double>>.generate(k, (_) => List<double>.filled(k, 0.0));
  final List<List<double>> atb =
      List<List<double>>.generate(k, (_) => List<double>.filled(3, 0.0));

  for (int i = 0; i < observed.length; i++) {
    final List<double> row = <double>[
      observed[i][0],
      observed[i][1],
      observed[i][2],
      1.0,
    ];
    for (int a = 0; a < k; a++) {
      for (int b = 0; b < k; b++) {
        ata[a][b] += row[a] * row[b];
      }
      for (int c = 0; c < 3; c++) {
        atb[a][c] += row[a] * expected[i][c];
      }
    }
  }

  final List<List<double>>? solution = _solve(ata, atb);
  if (solution == null) return null;

  return (
    matrix: <List<double>>[
      <double>[solution[0][0], solution[1][0], solution[2][0]],
      <double>[solution[0][1], solution[1][1], solution[2][1]],
      <double>[solution[0][2], solution[1][2], solution[2][2]],
    ],
    offset: <double>[solution[3][0], solution[3][1], solution[3][2]],
  );
}

/// Gaussian elimination with partial pivoting. Returns null if singular.
List<List<double>>? _solve(List<List<double>> a, List<List<double>> b) {
  final int n = a.length;
  final int m = b[0].length;
  final List<List<double>> matrix =
      List<List<double>>.generate(n, (int i) => <double>[...a[i], ...b[i]]);

  for (int col = 0; col < n; col++) {
    int pivot = col;
    for (int row = col + 1; row < n; row++) {
      if (matrix[row][col].abs() > matrix[pivot][col].abs()) pivot = row;
    }
    if (matrix[pivot][col].abs() < 1e-12) return null;
    final List<double> tmp = matrix[col];
    matrix[col] = matrix[pivot];
    matrix[pivot] = tmp;

    final double diagonal = matrix[col][col];
    for (int j = col; j < n + m; j++) {
      matrix[col][j] /= diagonal;
    }
    for (int row = 0; row < n; row++) {
      if (row == col) continue;
      final double factor = matrix[row][col];
      if (factor == 0) continue;
      for (int j = col; j < n + m; j++) {
        matrix[row][j] -= factor * matrix[col][j];
      }
    }
  }

  return List<List<double>>.generate(
    n,
    (int i) => matrix[i].sublist(n, n + m),
  );
}

List<List<double>> _toLinear(List<List<double>> rgbU8) {
  return rgbU8
      .map((List<double> c) => <double>[
            srgbToLinear(c[0] / 255.0),
            srgbToLinear(c[1] / 255.0),
            srgbToLinear(c[2] / 255.0),
          ],)
      .toList(growable: false);
}

ColorCorrection? fitCorrection(
  List<List<double>> observedU8,
  List<List<double>> expectedU8,
  String model,
) {
  final int n = observedU8.length;
  final int minimum = model == 'affine33' ? 4 : 2;
  if (n < minimum) return null;

  final List<List<double>> observedLinear = _toLinear(observedU8);
  final List<List<double>> expectedLinear = _toLinear(expectedU8);
  final List<Lab> expectedLab = expectedU8
      .map((List<double> c) => srgbU8ToLab(c[0], c[1], c[2]))
      .toList(growable: false);

  ({List<List<double>> matrix, List<double> offset})? fit;
  if (model == 'per_channel') {
    fit = _fitPerChannel(observedLinear, expectedLinear);
  } else {
    fit = _fitAffine33(observedLinear, expectedLinear);
  }
  if (fit == null) return null;

  List<double> apply(List<List<double>> m, List<double> o, List<double> v) => <double>[
        for (int r = 0; r < 3; r++)
          m[r][0] * v[0] + m[r][1] * v[1] + m[r][2] * v[2] + o[r],
      ];

  final List<double> residuals = <double>[
    for (int i = 0; i < n; i++)
      deltaE2000(
        _labOfLinear(apply(fit.matrix, fit.offset, observedLinear[i])),
        expectedLab[i],
      ),
  ];

  // Leave-one-out: refit without patch i, then measure the error on patch i.
  final List<double> cvResiduals = <double>[];
  if (n > minimum) {
    for (int i = 0; i < n; i++) {
      final List<List<double>> keepObserved = <List<double>>[
        for (int j = 0; j < n; j++)
          if (j != i) observedLinear[j],
      ];
      final List<List<double>> keepExpected = <List<double>>[
        for (int j = 0; j < n; j++)
          if (j != i) expectedLinear[j],
      ];
      final fold = model == 'per_channel'
          ? _fitPerChannel(keepObserved, keepExpected)
          : _fitAffine33(keepObserved, keepExpected);
      if (fold == null) {
        cvResiduals.add(residuals[i]);
        continue;
      }
      cvResiduals.add(
        deltaE2000(
          _labOfLinear(apply(fold.matrix, fold.offset, observedLinear[i])),
          expectedLab[i],
        ),
      );
    }
  } else {
    cvResiduals.addAll(residuals);
  }

  return ColorCorrection(
    model: model,
    matrix: fit.matrix,
    offset: fit.offset,
    residualsDe: residuals,
    cvResidualsDe: cvResiduals,
    nPatches: n,
  );
}

/// Fit every candidate model and keep the best cross-validated one.
ColorCorrection? fitBestCorrection(
  List<List<double>> observedU8,
  List<List<double>> expectedU8,
) {
  final List<ColorCorrection> fitted = <ColorCorrection>[
    for (final String model in <String>['per_channel', 'affine33'])
      if (fitCorrection(observedU8, expectedU8, model) case final ColorCorrection c) c,
  ];
  if (fitted.isEmpty) return null;
  fitted.sort((ColorCorrection a, ColorCorrection b) =>
      a.meanCvResidualDe.compareTo(b.meanCvResidualDe),);
  return fitted.first;
}

File: backend/app/services/colorimetry/correction.py
"""Colour correction from observed reference patches to true printed colour.

Model
-----
The correction is fitted in **linear** RGB, because that is the space in which a
camera's response to light is approximately affine. Fitting in gamma-encoded
sRGB would fold the transfer function into the fit and produce a correction that
is only valid at the exposure it was fitted at.

Two models are available:

``per_channel``
    Independent gain and offset per channel: ``out_c = g_c * in_c + o_c``.
    Six parameters, needs only the neutral ramp. This is the model spec 7
    describes ("compute a per-channel colour correction") and it handles the
    dominant error source, which is white-balance drift.

``affine33``
    A full 3x3 matrix plus offset, fitted by least squares over all patches.
    Twelve parameters. This additionally corrects channel cross-talk from the
    sensor's colour filter array, which per-channel gain cannot express.

The pipeline fits both and keeps whichever gives the lower **cross-validated**
residual. Cross-validation matters: the 3x3 model has enough freedom to fit 12
patches well while generalising worse to the strip colour, and an in-sample
residual would hide exactly that. Leave-one-out is cheap at this size and gives
an honest estimate of the error we are about to propagate into the confidence
interval.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

import numpy as np

from app.services.colorimetry.color_math import (
    linear_to_srgb,
    srgb_to_linear,
    srgb_u8_to_lab,
    xyz_to_lab,
)
from app.services.colorimetry.color_math import srgb_to_xyz as _srgb_to_xyz
from app.services.colorimetry.quality import residuals_between

CorrectionModel = Literal["per_channel", "affine33"]


@dataclass
class ColorCorrection:
    """A fitted correction plus the evidence that it is trustworthy."""

    model: CorrectionModel
    matrix: np.ndarray  # (3, 3)
    offset: np.ndarray  # (3,)
    residuals_de: np.ndarray  # in-sample per-patch CIEDE2000
    cv_residuals_de: np.ndarray  # leave-one-out per-patch CIEDE2000
    n_patches: int

    @property
    def mean_residual_de(self) -> float:
        return float(np.mean(self.residuals_de))

    @property
    def mean_cv_residual_de(self) -> float:
        return float(np.mean(self.cv_residuals_de))

    @property
    def worst_cv_residual_de(self) -> float:
        return float(np.max(self.cv_residuals_de))

    def apply_linear(self, linear_rgb: np.ndarray) -> np.ndarray:
        arr = np.atleast_2d(np.asarray(linear_rgb, dtype=np.float64))
        out = arr @ self.matrix.T + self.offset
        return out[0] if np.ndim(linear_rgb) == 1 else out

    def apply_srgb_u8(self, rgb_u8: np.ndarray) -> np.ndarray:
        """Correct an 8-bit sRGB colour, returning corrected sRGB in [0, 1]."""
        linear = srgb_to_linear(np.asarray(rgb_u8, dtype=np.float64) / 255.0)
        corrected = np.clip(self.apply_linear(linear), 0.0, 1.0)
        return linear_to_srgb(corrected)

    def apply_to_lab(self, rgb_u8: np.ndarray) -> np.ndarray:
        """Correct an 8-bit sRGB colour and return it as CIE L*a*b*."""
        corrected_srgb = np.atleast_2d(self.apply_srgb_u8(rgb_u8))
        lab = xyz_to_lab(_srgb_to_xyz(corrected_srgb))
        return lab[0] if np.ndim(rgb_u8) == 1 else lab

    def to_dict(self) -> dict:
        return {
            "model": self.model,
            "matrix": [[round(float(v), 6) for v in row] for row in self.matrix],
            "offset": [round(float(v), 6) for v in self.offset],
            "n_patches": self.n_patches,
            "mean_residual_de": round(self.mean_residual_de, 4),
            "mean_cv_residual_de": round(self.mean_cv_residual_de, 4),
            "worst_cv_residual_de": round(self.worst_cv_residual_de, 4),
        }


def _fit_per_channel(observed: np.ndarray, expected: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Least-squares gain and offset per channel, in linear RGB."""
    gains = np.ones(3)
    offsets = np.zeros(3)
    for c in range(3):
        design = np.column_stack([observed[:, c], np.ones(len(observed))])
        solution, *_ = np.linalg.lstsq(design, expected[:, c], rcond=None)
        gains[c], offsets[c] = solution
    return np.diag(gains), offsets


def _fit_affine33(observed: np.ndarray, expected: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Least-squares 3x3 matrix plus offset, in linear RGB."""
    design = np.column_stack([observed, np.ones(len(observed))])
    solution, *_ = np.linalg.lstsq(design, expected, rcond=None)  # (4, 3)
    return solution[:3].T, solution[3]


_FITTERS = {"per_channel": _fit_per_channel, "affine33": _fit_affine33}


def _lab_of_linear(linear_rgb: np.ndarray) -> np.ndarray:
    srgb = linear_to_srgb(np.clip(linear_rgb, 0.0, 1.0))
    return xyz_to_lab(_srgb_to_xyz(np.atleast_2d(srgb)))


def fit_correction(
    observed_rgb_u8: np.ndarray,
    expected_rgb_u8: np.ndarray,
    model: CorrectionModel = "per_channel",
) -> ColorCorrection:
    """Fit one correction model and evaluate it, including leave-one-out.

    Parameters
    ----------
    observed_rgb_u8
        (N, 3) means measured off the photographed card.
    expected_rgb_u8
        (N, 3) the values those patches are printed to.
    """
    observed_u8 = np.asarray(observed_rgb_u8, dtype=np.float64).reshape(-1, 3)
    expected_u8 = np.asarray(expected_rgb_u8, dtype=np.float64).reshape(-1, 3)
    n = len(observed_u8)

    minimum = 4 if model == "affine33" else 2
    if n < minimum:
        raise ValueError(f"{model} needs at least {minimum} patches, got {n}")

    obs_linear = srgb_to_linear(observed_u8 / 255.0)
    exp_linear = srgb_to_linear(expected_u8 / 255.0)
    expected_lab = srgb_u8_to_lab(expected_u8)

    matrix, offset = _FITTERS[model](obs_linear, exp_linear)

    corrected = obs_linear @ matrix.T + offset
    residuals = residuals_between(_lab_of_linear(corrected), expected_lab)

    # Leave-one-out: refit without patch i, then measure the error on patch i.
    cv_residuals = np.zeros(n)
    if n > minimum:
        for i in range(n):
            keep = np.arange(n) != i
            m_i, o_i = _FITTERS[model](obs_linear[keep], exp_linear[keep])
            predicted = obs_linear[i] @ m_i.T + o_i
            cv_residuals[i] = float(
                residuals_between(_lab_of_linear(predicted), expected_lab[i : i + 1])[0]
            )
    else:
        # Not enough patches to hold one out; fall back to the in-sample value
        # and let the caller's threshold decide. This should not occur with the
        # v1 card, which has 12 patches.
        cv_residuals = residuals.copy()

    return ColorCorrection(
        model=model,
        matrix=matrix,
        offset=offset,
        residuals_de=residuals,
        cv_residuals_de=cv_residuals,
        n_patches=n,
    )


def fit_best_correction(
    observed_rgb_u8: np.ndarray,
    expected_rgb_u8: np.ndarray,
    candidates: tuple[CorrectionModel, ...] = ("per_channel", "affine33"),
) -> ColorCorrection:
    """Fit every candidate model and keep the best cross-validated one."""
    fitted: list[ColorCorrection] = []
    for model in candidates:
        try:
            fitted.append(fit_correction(observed_rgb_u8, expected_rgb_u8, model))
        except (ValueError, np.linalg.LinAlgError):
            continue
    if not fitted:
        raise ValueError("no correction model could be fitted to the observed patches")
    return min(fitted, key=lambda c: c.mean_cv_residual_de)
